core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -237,13 +227,12 @@ public synchronized IndexShard createShard(int sShardId, ShardRouting routing) {
         }
         final Settings indexSettings = settingsService.getSettings();
         final ShardId shardId = new ShardId(index, sShardId);
-        ShardLock lock = null;
+        final ShardLock lock = nodeEnv.shardLock(shardId, TimeUnit.SECONDS.toMillis(5));
```

Comment:
why did you take this out of the try catch block? I'm fine with keeping this as is, but then the catch IOException clause is probably redundant:

```
catch (IOException e) {
             ElasticsearchException ex = new ElasticsearchException("failed to create shard", e);
             ex.setShard(shardId);
             throw ex;
         }
```